File: api/src/classes/units/Sniper.js
const Infantry = require('./infantry')

class Sniper extends Infantry {
  constructor(square, team) {
    super(square, team)
    this.type = 'sniper'
  }

  assignThreat() {
    super()
    this.square
      .neighbors({
        headings: [[0, 1], [1, 1]],
        repeat: Infinity,
        test: square => !square.isObstructed,
        inclusive: true
      })
      .filter(square => square.isPassable)
      .forEach(square => this.threaten(square))
  }
}

module.exports = Sniper

File: api/lib/units/sniper.rb
require_relative './infantry.rb'

class Sniper < Infantry
  def initialize square, team
    super
    @type = :sniper
  end

  def assignThreat
    super
    @square
      .neighbors(
        headings: [[0, 1], [1, 1]],
        repeat: Float::INFINITY,
        test: proc { |square| !square.obstructed? },
        inclusive: true
      )
      .each { |square| threaten square if square.passable?}
  end
end